### ISSUE TITLE
[ExpoKit IOS] Background location task deleted when app killed

### ISSUE THREAD

**Issue by narefyev91** [expo/expo#4860]
### Environment

 Expo CLI 2.20.8 environment info:
    System:
      OS: macOS 10.14.5
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 10.14.2 - /usr/local/bin/node
      Yarn: 1.16.0 - /usr/local/bin/yarn
      npm: 6.9.0 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    IDEs:
      Android Studio: 3.4 AI-183.6156.11.34.5522156
      Xcode: 10.2.1/10E1001 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^33.0.0 => 33.0.3 
      react: 16.8.3 => 16.8.3 
      react-native: https://github.com/expo/react-native/archive/sdk-33.0.0.tar.gz => 0.59.8 
      react-navigation: ^3.11.0 => 3.11.0 
    npmGlobalPackages:
      expo-cli: 2.20.8
### Steps to Reproduce
When the app in foreground and background - location tracked and stored in database, when the app killed - background task killed - no request is going in database.

### Expected Behavior
Location should be stored also when the app is killed

### Actual Behavior
I don't think I missed some additional options - because all works in background. I also know that for iOS it's needed to walk more than 500m. Also one question - what could be defined in background task when it's killed? Could we use AsyncStorage, or simple fetch to send data to database or using moment or axios? Are there any limits? Because previously when I didn't eject - all this logic was working fine... Any ideas and helps will be really appreciated

**Comment by cruzach** [comment]
This may be a duplicate of [#3503](https://github.com/expo/expo/issues/3503), although that was standalone and this is Expokit.

**Comment by mikaarefyev** [comment]
I think no, it something specific - I add some updates - after moving 500m, I saw that line which indicate that background location starts to track appears, but after one minute it disappears with location icon - please see attached images. 
https://ibb.co/WpR3S3x
https://ibb.co/8BbMXG3
In this case - again my question - maybe when significant location change - and ios rise some js code to implement - we can't use all futures such as async storage, fetch methods and etc.?

**Comment by narefyev91** [comment]
@tsapeta Hey any ideas about it?

**Comment by narefyev91** [comment]
Also find solution to let this work on latest IOS for basic expo app, but not for ejected app - For ejected it delete this task anyway.
Also one question for @tsapeta and @cruzach I try to use other library https://github.com/transistorsoft/react-native-background-geolocation to let my problem resolve - but also this package doesn't work when app killed, but on clean react-native app it's works and I think the problem is inside expo task manager - because this package also use manager to launch app in background and do some code... Maybe you see the same problem previously - and you have any ideas how to deal with it? Thanks

**Comment by dereknelson** [comment]
@narefyev91 fwiw, I tried both rn-bgloc and radar, and basically they're incompatible with expo. [here's the conclusion we came to](https://github.com/radarlabs/react-native-radar/issues/77#issuecomment-484218339)

I was able to use both async storage and fetch, however keep in mind you have ~5 seconds for your code to run on ios before the system shuts it down, so if you're doing multiple async fetch calls, depending on response time you may not get them all done.

**Comment by k2xl** [comment]
Did anyone find a workaround? Background location definitely not working on Android when app is killed. Testing via CURLing as well as using notification

**Comment by k2xl** [comment]
Also, just for more details (for Android). Just tried on SDK 34 and still no dice.

I tried add `foregroundService` as an option to `startLocationUpdatesAsync` and then a timeInterval of 10000 (10 seconds)

Something very interesting results. So basically if I minimize the app (hit home button and just go to another screen) the service stays up and polls at same interval (tested with curl and local push notifications).

However, if I swipe up on the app from the recent apps screen then here's what happens. The foreground service stays visible... until the next `tick` for the background service (in my test case 10 seconds) and then it "crashes"

If I increase to 30 seconds and remove the app as soon as a tick happens I wait about 30 seconds then the app "crashes" and the foreground service goes away.

I say "crashes" though no crash warning pops up.



I managed to grab output of `adb logcat`
```
07-30 23:53:46.719  7673  7673 I TaskService: Handling job with task name 'background-location-updates' for app with ID '<APP ID>'.
07-30 23:53:46.720  7673  7673 E Expo    : Cannot initialize app loader. <init> [class android.content.Context]
07-30 23:53:46.720  7673  7673 E TaskService: Cannot execute background task because application loader can't be found.
07-30 23:53:46.720  7673  7673 I TaskService: Unregistering task 'background-location-updates' for app '<APP ID>'.
07-30 23:53:46.734  1150  1616 I GnssLocationProvider: WakeLock acquired by sendMessage(SET_REQUEST, 0, com.android.server.location.GnssLocationProvider$GpsRequest@b4ea926)
07-30 23:53:46.736  1150  1172 I GnssLocationProvider: WakeLock released by handleMessage(SET_REQUEST, 0, com.android.server.location.GnssLocationProvider$GpsRequest@b4ea926)
07-30 23:53:46.739  7673  7673 I TaskService: Handling intent with task name 'background-location-updates' and appId '<APP ID>'.
07-30 23:53:46.739  7673  7673 W TaskService: Task or consumer not found.
```

Seems the first error is `Cannot initialize app loader. <init> [class android.content.Context]` from Expo.

**Comment by zsayenko** [comment]
Confirming that geofencing isn't working either when the app is killed on iOS.
Everything works correctly when the app is in the background, but if I swipe up the app and kill it completely geofencing stops working.

Any updates on this would be greatly appreciated.

**Comment by tip3x** [comment]
On IOS: 
Background location worked as expected even after the app was killed (app waked up after 500m+ change from the last location).

On Android:
Tried everything, it just didn't work after the app killed. The foregroundService flag didn't seem to do anything. Tried it in both expo client and as a standalone app.

**Comment by nivniv850** [comment]
happened to me also like @tip3x described. 
Can someone fix it ?

**Comment by k2xl** [comment]
@tsapeta can you confirm this bug's visibility with expo team? is this something planned on being fixed?

**Comment by dvester** [comment]
Really need some sort of resolution on this. Background location is not working when the app is closed. Can a member of the expo team please respond.

**Comment by glauciowathier** [comment]
@dvester, @k2xl, @nivniv850, @tip3x
Did any of you get around the problem?

**Comment by dvester** [comment]
No, no resolution on this. We had the shelve our background tracking code because of this. The unreliability of it, isn't something we want included in our product.

**Comment by glauciowathier** [comment]
@dvester thanks for your reply. I need to found another way to solve this, because it's indispensable in my project, maybe I will need to eject my app

**Comment by niv-breez** [comment]
I tried to eject the app but it still didn't worked, because expo wrapped things in the native code.
It worked only after I create a new pure react native app without expo, and moved my code manually from the ejected expo project to the new project.
I am using external library for location services and it worked in the background and foreground on IOS and on Android.

**Comment by glauciowathier** [comment]
@tsapeta @cruzach is there anything new about this? Sorry for the insistence.

**Comment by tompap** [comment]
@niv-breez which external library did you use ? to make it work on a classical react native app ?

**Comment by niv-breez** [comment]
https://github.com/transistorsoft/react-native-background-geolocation

But take a look also on:
https://radar.io/
I found it just after I integrated the transistorsoft plugin

**Comment by zsayenko** [comment]
Hey guys, any updates on this? It's been a while

**Comment by k2xl** [comment]
@glauciowathier @dvester @zsayenko - No updates that I could see. Unsure if Expo team has acknowledged the issue yet. Still not working for me with Expo 35.

**Comment by michaeldiazpinion** [comment]
Any updates?

**Comment by MGB247** [comment]
I would really like a solution. Our app is only held back by this. It's like the last thing we want before production.

**Comment by maguayo** [comment]
Any updates?

**Comment by mustaqeem** [comment]
An update on this?

**Comment by VitorValandro** [comment]
Hey guys, any update? Still not working for me...

**Comment by nimrodbens** [comment]
any update? not working with `expo-location` 9.0.1 and `expo-task-manager` 8.5.0.

i'm on a bare project.

**Comment by maguayo** [comment]
Not solved yet

**Comment by macsint** [comment]
Why has this been closed since it has not been solved?

**Comment by macsint** [comment]
@tsapeta the problem here is that from what I can dig in about 2 days that there are numerous issues linking back to the same problem which is that the background location is not getting updated through a task after the app is killed.

It seems to be related to the fact that, according to the [Expo Documentation](https://docs.expo.dev/versions/latest/sdk/background-fetch/): `BackgroundFetch only works when the app is backgrounded, not if the app was terminated or upon device reboot. You can check out [the relevant GitHub issue](https://github.com/expo/expo/issues/3582) for more details.` which I'm assuming is what ultimately leads to this issue. 

Other issues which are acknowledging this problem get closed without a resolution and so is there a reason why we are not addressing this? 

Here's [another example](https://github.com/expo/expo/issues/3582) of an issue that is directly mentioned in the Expo docs for background fetch which got closed.

**Comment by tsapeta** [comment]
What do you mean by "the app is killed"? I think there is some misunderstanding – it's the platform that doesn't allow to execute any background fetch or silent push notifications (probably location as well) when the user killed the app from multitasking UI.

**Comment by macsint** [comment]
So what I mean is that both background fetch and the expo location packages do not seem to support providing updates and/or executing code in iOS specifically once the app is killed/terminated. They work while the app is in the Foreground or Background, but not once the app is manually closed by the user.

According to your documentation it seems like background fetch supports bringing up the app live in the background once it was closed on Android only, and thus enabling us to execute certain tasks even after the app was closed.

When you say

**Comment by macsint** [comment]
@tsapeta I created an issue for this since it is an actual issue, and I have further tested this. It is recognized as an issue in your documentation by your team and seems to be a big blocker for a lot of people based on the issues that were open without resolution and/or all the StackOverFlow mentioned, although not sure why this has not been addressed.

Since this is a big blocker for my team, I am volunteering to contribute towards fixing this if possible.

Here's the issue that I created: #27933 27933
